resallocserver/app.py

```diff
@@ -15,25 +15,51 @@
 # with this program; if not, write to the Free Software Foundation, Inc.,
 # 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 
-import os
+from contextlib import contextmanager
+
 from sqlalchemy import create_engine
 from sqlalchemy.pool import NullPool
 from sqlalchemy.orm import sessionmaker, scoped_session
-from contextlib import contextmanager
 
-from resallocserver.config import CONFIG
+from resallocserver.config import get_config
+from resallocserver.log import get_logger
+
+class AppContext:
+    """
+    Provide a singleton-like behavior for object attributes.  One needs to
+    define instantiate_* methods.
+    """
+    # pylint: disable=missing-function-docstring
+    def __getattr__(self, key):
+        if key.startswith("instantiate_"):
+            raise KeyError("Please define {} method in AppContext".format(key))
+        setattr(self, key, getattr(self, "instantiate_{}".format(key))())
+        return getattr(self, key)
+
+    @staticmethod
+    def instantiate_config():
+        return get_config()
+
+    @staticmethod
+    def instantiate_log():
+        return get_logger("resallocserver")
+
+    def instantiate_engine(self):
+        return create_engine(self.config['db_url'], poolclass=NullPool)
+
+    def instantiate_Session(self):  # pylint: disable=invalid-name
```

Comment:
I know it is intentional but the capital S is weird.

Reply:
Indeed, we can live with small S, even though SQLAlchemy has it differently.